XenAdmin/Commands/NewTemplateFromSnapshotCommand.cs

```diff
@@ -82,34 +86,38 @@ protected override void RunCore(SelectedItemCollection selection)
             {
                 if (dialog.ShowDialog(Parent) == DialogResult.OK)
                 {
-                    // TODO: work out what the new description should be
-                    var action = new VMCloneAction(snapshot, dialog.InputText, "");
+                    var action = new VMCloneAction(snapshot, dialog.InputText, string.Format(Messages.TEMPLATE_FROM_SNAPSHOT_DEFAULT_DESCRIPTION, BrandManager.BrandConsole, snapshot.Name()));
                     action.Completed += action_Completed;
                     action.RunAsync();
                 }
             }
         }
 
-        void action_Completed(ActionBase sender)
+        private static void action_Completed(ActionBase sender)
         {
-            AsyncAction action = (AsyncAction)sender;
-            var vm = action.Connection.Resolve(new XenRef<VM>(action.Result));
+            var action = (AsyncAction)sender;
+            VM vm = null;
+            try

```

Comment:
This is one of the cases where the exception does not need catching, but rather prevention.
When the action is cancelled or fails, the getter of action.Result throws that cancellation or failure exception (it's debatable whether this is a good design, but this is a different story), so we shoud not try to get the Result unless the action has succeeded. Instead of the try-catch, all this handler needs is a check at the beginning:
```
if (!sender.Succeeded || !(sender is AsyncAction action))
    return;
```
I've seen this once before, generally we do check before getting the Result, but there might be some forgotten case somewhere, an audit might be a good idea.